### ISSUE TITLE
Take care about missing or misconcepted energy-consumption notarizaitons on RDDL by the validators.

### ISSUE THREAD

**Issue by eckelj** [rddl-network/rddl-info#22]
Rddl-info does not show that all validators notarize their energy consumptions accordingly. This is something that needs to be fixed.

**Comment by eckelj** [comment]
the following nodes do not notarize energy consumption:

node1, node2, node4, node7

I investigated node 1 and found out that the data was notarized at  (grep
`Mar 13 15:46:34 raspberrypi poetry[1497]: INFO:     192.168.0.88:56614 - "POST /assets?cid=bafkreif6xyusum4o77k6t4s6znzvu7dpo7ah7ucalmijql3474fw62zguu HTTP/1.1" 200 OK`

the service was showing me that it was running since two weeks:
```
rddl@raspberrypi:~ $ sudo systemctl status 0x21e8.service
● 0x21e8.service - RDDL Attestation Service
     Loaded: loaded (/etc/systemd/system/0x21e8.service; enabled; vendor preset: enabled)
     Active: active (running) since Tue 2023-03-07 07:17:07 GMT; 2 weeks 1 days ago
       Docs: https://rddl.io
```
The API was not reachable and there wasn't any reaction on calling `curl http://localhost:8000/docs`

Interestingly, the process details showed that the uivcorn process got started: on the 14th of March, one day after the last valid activity:

```
rddl@raspberrypi:~ $ ps -Af | grep uvicorn
rddl       376     1 14 Mar07 ?        2-05:49:49 /home/rddl/.cache/pypoetry/virtualenvs/x21e8-6NXnCYQ5-py3.9/bin/python /home/rddl/.cache/pypoetry/virtualenvs/x21e8-6NXnCYQ5-py3.9/bin/uvicorn --log-level debug --reload x21e8.main:app --host 0.0.0.0

```
At the same time, the systemd didn't recognize a downtime. I had to kill the process. systemctl restart didn't respond/work as expected.

**Comment by eckelj** [comment]
@jmastr I restarted the service on node1 and it's working there again. I assume that the other nodes have the same issue. can you be so kind as to investigate?

**Comment by jmastr** [comment]
On node4 the last successful message was on Mar 13:
```
Mar 13 15:45:21 raspberrypi poetry[511]: INFO:     192.168.0.39:37466 - "POST /assets?cid=bafkreigwwgdmni5rp375ha4jgp5yjiy74ekytnw47hkb5rcwfwvke5hw5a HTTP/1.1" 200 OK
```

Only a small part of it succeeded:
```
rddl@raspberrypi:~ $ sudo journalctl |grep "POST /assets"|grep "200 OK"|wc -l
1395
rddl@raspberrypi:~ $ sudo journalctl |grep "POST /assets"|grep "423 Locked"|wc -l
6183
rddl@raspberrypi:~ $ sudo journalctl |grep "POST /assets"|grep "500 Internal Server Error"|wc -l
623
```

Here one more time 0x21e8 is stuck:
```
rddl@raspberrypi:~ $ sudo systemctl status 0x21e8.service
● 0x21e8.service - RDDL Attestation Service
     Loaded: loaded (/etc/systemd/system/0x21e8.service; enabled; vendor preset: enabled)
     Active: active (running) since Thu 2023-02-02 08:59:04 GMT; 1 months 17 days ago
       Docs: https://rddl.io
   Main PID: 344 (uvicorn)
      Tasks: 3 (limit: 1593)
        CPU: 3d 1h 4min 2.593s
     CGroup: /system.slice/0x21e8.service
             ├─344 /home/pi/.cache/pypoetry/virtualenvs/x21e8-nw7avRiT-py3.9/bin/python /home/pi/.cache/pypoetry/virtualenvs/x21e8-nw7avRiT-py3.9/bin/uvicorn --log-level debug --reload x21e8.main:app --host 0.0.0.0
             ├─510 /home/pi/.cache/pypoetry/virtualenvs/x21e8-nw7avRiT-py3.9/bin/python -c from multiprocessing.resource_tracker import main;main(4)
             └─511 /home/pi/.cache/pypoetry/virtualenvs/x21e8-nw7avRiT-py3.9/bin/python -c from multiprocessing.spawn import spawn_main; spawn_main(tracker_fd=5, pipe_handle=7) --multiprocessing-fork

Mar 13 15:15:07 raspberrypi poetry[511]: b'\x84\x12[\x0b*V^\x92r\xc8l\xa5\xd9%a-\x8f\xec<\x1e\x07V\xf5\xe4\x99\x8a\xdb\xf8\xebX\xc5\x9b'
Mar 13 15:15:07 raspberrypi poetry[511]: INFO:     192.168.0.39:35594 - "POST /assets?cid=bafkreid4ex5z2hyvsjz6oi3mjkgqcipypfclg4z5ooncqh45xgych5fble HTTP/1.1" 200 OK
Mar 13 15:31:34 raspberrypi poetry[511]: INFO:     192.168.0.39:55186 - "POST /data?encrypt=False HTTP/1.1" 200 OK
Mar 13 15:31:36 raspberrypi poetry[511]: b'\x84\x12[\x0b*V^\x92r\xc8l\xa5\xd9%a-\x8f\xec<\x1e\x07V\xf5\xe4\x99\x8a\xdb\xf8\xebX\xc5\x9b'
Mar 13 15:31:36 raspberrypi poetry[511]: b'\x84\x12[\x0b*V^\x92r\xc8l\xa5\xd9%a-\x8f\xec<\x1e\x07V\xf5\xe4\x99\x8a\xdb\xf8\xebX\xc5\x9b'
Mar 13 15:31:36 raspberrypi poetry[511]: INFO:     192.168.0.39:52564 - "POST /assets?cid=bafkreigj36qf5jw5nxvdy45riazbcsumfsz5ybwatuwdurnh2roqtflwse HTTP/1.1" 200 OK
Mar 13 15:45:20 raspberrypi poetry[511]: INFO:     192.168.0.39:35114 - "POST /data?encrypt=False HTTP/1.1" 200 OK
Mar 13 15:45:21 raspberrypi poetry[511]: b'\x84\x12[\x0b*V^\x92r\xc8l\xa5\xd9%a-\x8f\xec<\x1e\x07V\xf5\xe4\x99\x8a\xdb\xf8\xebX\xc5\x9b'
Mar 13 15:45:21 raspberrypi poetry[511]: b'\x84\x12[\x0b*V^\x92r\xc8l\xa5\xd9%a-\x8f\xec<\x1e\x07V\xf5\xe4\x99\x8a\xdb\xf8\xebX\xc5\x9b'
Mar 13 15:45:21 raspberrypi poetry[511]: INFO:     192.168.0.39:37466 - "POST /assets?cid=bafkreigwwgdmni5rp375ha4jgp5yjiy74ekytnw47hkb5rcwfwvke5hw5a HTTP/1.1" 200 OK
```

**Comment by jmastr** [comment]
> At the same time, the systemd didn't recognize a downtime. I had to kill the process. systemctl restart didn't respond/work as expected.

Again, I suspect the virtualenv. I guess it give systemd a hard time. Can we run it "natively"?

**Comment by jmastr** [comment]
I cannot reproduce the above error. What I can do is reproduce this behavior, which I think could be similar. The main "issue" is that we use `uvicorn`, which is more for development and does not handle if the worker process exits. Have a look at PID 12310:
```
$ sudo systemctl status 0x21e8.service                                                                                                                                                                                                                                                                                                     
● 0x21e8.service - RDDL Attestation Service                                                                                                                                           
     Loaded: loaded (/etc/systemd/system/0x21e8.service; enabled; vendor preset: enabled)                                                                                                                                                                                                                                                                                   
     Active: active (running) since Wed 2023-05-31 08:12:53 BST; 6s ago                                                                                                                                                                                                                                                                                                     
       Docs: https://rddl.io                                                                                                                                                          
   Main PID: 12305 (uvicorn)                                                                                                                                                          
      Tasks: 3 (limit: 472)                                                                                                                                                           
        CPU: 7.996s                                                                                                                                                                   
     CGroup: /system.slice/0x21e8.service                                                                                                                                             
             ├─12305 /home/rddl/0x21e8/venv/bin/python3 /home/rddl/0x21e8/venv/bin/uvicorn --log-level debug --reload x21e8.main:app --host 0.0.0.0                                   
             ├─12309 /home/rddl/0x21e8/venv/bin/python3 -c from multiprocessing.resource_tracker import main;main(4)                                                                  
             └─12310 /home/rddl/0x21e8/venv/bin/python3 -c from multiprocessing.spawn import spawn_main; spawn_main(tracker_fd=5, pipe_handle=7) --multiprocessing-fork                                                                                                                                                                                                     
                                                                                                                                                                                      
May 31 08:12:53 raspberrypi systemd[1]: Started RDDL Attestation Service.                                                                                                             
May 31 08:12:53 raspberrypi uvicorn[12305]: INFO:     Will watch for changes in these directories: ['/home/rddl/0x21e8']                                                              
May 31 08:12:53 raspberrypi uvicorn[12305]: INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)                                                                   
May 31 08:12:53 raspberrypi uvicorn[12305]: INFO:     Started reloader process [12305] using StatReload                                                                               
May 31 08:12:56 raspberrypi uvicorn[12310]: INFO:     Started server process [12310]                                                                                                  
May 31 08:12:56 raspberrypi uvicorn[12310]: INFO:     Waiting for application startup.                                                                                                
May 31 08:12:56 raspberrypi uvicorn[12310]: INFO:     Application startup complete.
```

Due to a problem with `libwally-core`, that we have in the past, I can crash that process and `uvicorn` will neither recover nor restart the systemd unit, because the main process is still running:
```
$ sudo systemctl status 0x21e8.service 
● 0x21e8.service - RDDL Attestation Service
     Loaded: loaded (/etc/systemd/system/0x21e8.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2023-05-31 08:12:53 BST; 2min 12s ago
       Docs: https://rddl.io
   Main PID: 12305 (uvicorn)
      Tasks: 3 (limit: 472)
        CPU: 1min 53.753s
     CGroup: /system.slice/0x21e8.service
             ├─12305 /home/rddl/0x21e8/venv/bin/python3 /home/rddl/0x21e8/venv/bin/uvicorn --log-level debug --reload x21e8.main:app --host 0.0.0.0
             └─12309 /home/rddl/0x21e8/venv/bin/python3 -c from multiprocessing.resource_tracker import main;main(4)

May 31 08:12:53 raspberrypi systemd[1]: Started RDDL Attestation Service.
May 31 08:12:53 raspberrypi uvicorn[12305]: INFO:     Will watch for changes in these directories: ['/home/rddl/0x21e8']
May 31 08:12:53 raspberrypi uvicorn[12305]: INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
May 31 08:12:53 raspberrypi uvicorn[12305]: INFO:     Started reloader process [12305] using StatReload
May 31 08:12:56 raspberrypi uvicorn[12310]: INFO:     Started server process [12310]
May 31 08:12:56 raspberrypi uvicorn[12310]: INFO:     Waiting for application startup.
May 31 08:12:56 raspberrypi uvicorn[12310]: INFO:     Application startup complete.
May 31 08:13:30 raspberrypi uvicorn[12310]: INFO:     192.168.0.39:47048 - "POST /data?encrypt=False HTTP/1.1"
```

Here a quote:
> Up to this point, with all the tutorials in the docs, you have probably been running a server program like Uvicorn, running a single process.

> When deploying applications you will probably want to have some replication of processes to take advantage of multiple cores and to be able to handle more requests.

> ...

> Gunicorn is mainly an application server using the WSGI standard. That means that Gunicorn can serve applications like Flask and Django. Gunicorn by itself is not compatible with FastAPI, as FastAPI uses the newest [ASGI standard](https://asgi.readthedocs.io/en/latest/).

> But Gunicorn supports working as a process manager and allowing users to tell it which specific worker process class to use. Then Gunicorn would start one or more worker processes using that class.

> And Uvicorn has a Gunicorn-compatible worker class.

https://fastapi.tiangolo.com/deployment/server-workers/

Long story short I suggest to switch from `uvicorn` to `gunicorn` for the validator deployment.
